Analyze table function arguments #4

mohsaka opened the pull request:
Changes adapted from trino/PR#13602


## Description
<!---Describe your changes in detail-->

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* ... 
* ... 

Hive Connector Changes
* ... 
* ... 
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

mohsaka committed:
Rename AST nodes to avoid conflicting class names

Changes adapted from trino/PR#13602
Original commit: 21695a1dfb33d5cb8bb195397d769222da5a6345
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Fix table function argument resolution

Changes adapted from trino/PR#13602
Original commit: 8be4a486a525056ef6b361618280ae0185d7211b
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Refactor empty table treatment into a Node and Fix aliasing table arguments of table functions

Changes adapted from trino/PR#13649
Changes adapted from trino/PR#13602
Original commit: a63cc9c429cea7849cf723598719812cc8e5636c
Original commit: 31b2b79751944eed42eaa4538a52dbf65c0399a6
Author: kasiafi

Modifications were made to adapt to Presto including:
Updated tests to match Presto format.

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka commented:
Completed up to 31b2b79751944eed42eaa4538a52dbf65c0399a6

mohsaka committed:
Extract method for analyzing scalar arguments

Changes adapted from trino/PR#13602
Original commit: a93597ff72d98d577b32f5da6ef562ec85a7aad6
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Use variable

Changes adapted from trino/PR#13602
Original commit: 21acef5d028565b919e3eae2df9784fbc2317ae9
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Enforce the prune when empty property in TableArgumentSpecification

Prune when empty is enforced for a table argument with row semantics.
For a table argument with set semantocs, keep when empty is the default,
and it can be changed to prune when empty.

Changes adapted from trino/PR#13602
Original commit: 9771edb626ea47f2333988178107c7a2dc46ec6b
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Add comment

Changes adapted from trino/PR#13602
Original commit: 15664c845f4958ecffbb0dca2596226ee899c12e
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Simplify table argument representation in SPI

According to the SQL standard, the only properties
of a table argument used during analysis are: the row type,
partitioning, and ordering. Other information, like the
prune / keep when empty property, are not involved in the
analysis.

Changes adapted from trino/PR#13602
Original commit: 18e97d406ae89528de5d04c868587f3c6bc2f1c4
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Record relation names during analysis

Changes adapted from trino/PR#13602
Original commit: 866f83c965500bfc962113db7c65b4c8a7931d0b
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Remove descriptor mapping from table function analysis and planning

This is a change of approach on resolving arguments
of table functions.
Before this change, the mapping of descriptor arguments
to table arguments had to be determined during query
analysis. It was too limiting: in that model, input table
references were limited to the specified descriptor fields.

After this change, any input fields can be referenced,
and the required fields (channels) are to be determined
dynamically during function compilation.

One advantage of determining the required fields early
was the possibility to prune any columns that were
not required by the table function. This should be replaced
by adding a way for a table function to declare required
fields during query optimization.

Changes adapted from trino/PR#13602
Original commit: cbde4d7168efddf1d56f906d3d1e50d542d2f851
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Use Optional.orElseThrow()

Changes adapted from trino/PR#13602
Original commit: dc3552f812be96211989b73d60f60a99ab943863
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Add equals() and hashCode() methods for descriptor argument

It should make using descriptor arguments easier for the
table function author, and allow to compare with
NULL_DESCRIPTOR by equality.

Changes adapted from trino/PR#13602
Original commit: 612ad9bc1a5a3c5b80d5e5aacf7d5f3529a11a19
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Add serialization of table function scalar arguments

Changes adapted from trino/PR#13602
Original commit: 8568ed11b94130c2a76ad1cf99c04d91fd2a0eca
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka committed:
Copy a map argument in the constructor

Changes adapted from trino/PR#13602
Original commit: b7460da53871b9b6a77ed08c24afbee1d5f72dd9
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka commented:
Latest update, almost finished the last commit. Needs to bring in the Statement Analyzer into AbstractAnalyzer test. Afterwards we can work test by test to fix the issues that occur.

xin-zhang2 committed:
Fix error messages as we do not have NodeLocations associated with Qualified Names
Modify error messages to use Types instead of DataType from Trino
Fix test cases to follow Presto syntax.

Co-authored-by: mohsaka <135669458+mohsaka@users.noreply.github.com>

xin-zhang2 commented:
Lastest update:
Adde the table functions in AbstractAnalyzerTest.
Fixed some failures in the tests.

The current failure is in `testDescriptorArgument`
```
java.lang.AssertionError: Expected error 'line 1:80: Unknown type: verybigint', but got 'verybigint'

	at org.testng.Assert.fail(Assert.java:98)
	at com.facebook.presto.sql.analyzer.AbstractAnalyzerTest.assertFails(AbstractAnalyzerTest.java:473)
	at com.facebook.presto.sql.analyzer.AbstractAnalyzerTest.assertFails(AbstractAnalyzerTest.java:425)
	at com.facebook.presto.sql.analyzer.TestAnalyzer.testDescriptorArgument(TestAnalyzer.java:2076)

Caused by: com.facebook.presto.sql.analyzer.SemanticException: verybigint
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.lambda$null$30(StatementAnalyzer.java:1594)
	at java.util.Optional.map(Optional.java:215)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.lambda$null$31(StatementAnalyzer.java:1589)
	...
```
It is caused by a signature mismatch of SemanticException, and the code lies in
https://github.com/mohsaka/presto/blob/699b0ad948bf91b58a7955cbf54dd26837c3dd23/presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java#L1589-L1595

where `type` is a String in Presto but a `DataType` which extends `Node` in Trino. 

The same issue also occurs in`testCopartitioning`
https://github.com/mohsaka/presto/blob/699b0ad948bf91b58a7955cbf54dd26837c3dd23/presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java#L1698-L1702 
where `originalParts` is a list of String in Presto but a list of Identifier in Trino.

In order to indicate the accurate location in the error message, I think we should choose the trino approach and refactor the `type` field in `DescriptorField` and `originalParts` in `QualifiedName`
@mohsaka What's your thought on this?

mohsaka committed:
Analyze table and descriptor arguments

Changes adapted from trino/PR#13602
Original commit: dbef4d9be37494967496230573ab400e54aab0d9
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>
Co-authored-by: Xin Zhang <desertsxin@gmail.com>

mohsaka committed:
Validate return type declaration of table function at registration

Changes adapted from trino/PR#13653
Original commit: 4a0188ff8bab16bfd8f81449d12d7e7d96676f1c
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

mohsaka commented:
Update, started 13653 on the same PR since it is small. Will add 14115 to this PR as well since it is also small.

xin-zhang2 committed:
Analyze the context of table function invocation

Changes adapted from trino/PR#13653
Original commit: 4827fa8d76ac51706b47c108cc25bdd3bc3b5767
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

xin-zhang2 committed:
Analyze aliasing of table function invocation

Changes adapted from trino/PR#13653
Original commit: 02f3583d231fd7a89a18f48a6bfbec44b981c829
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

xin-zhang2 committed:
Fix test cases to follow Presto syntax.
Comment out the MATCH_RECOGNIZE test cases.

xin-zhang2 committed:
Add columns from input tables to table function returned relation type

Changes adapted from trino/PR#14115
Original commit: 6e95fb186fc3edc7fd6a5bdc385e68f8accff11e
Author: kasiafi

Co-authored-by: kasiafi <30203062+kasiafi@users.noreply.github.com>

xin-zhang2 committed:
Fix test cases to follow Presto syntax.
Comment out the MATCH_RECOGNIZE test cases.

xin-zhang2 committed:
Fix format issues.
